### ISSUE TITLE
Kernel Protocol Documentation & Third Party Kernel Registry

### ISSUE THREAD

**Issue by aaronclong** [polynote/polynote#1159]
I've been messing around with your notebook and love it!

Do you have a documented kernel spec for adding languages to the notebook? I want to make an experimental one using Swift that probably shouldn't be in the main repository. I was looking for more than just examples but also documentation.

Since the application uses the JVM, it could load a library from Maven Central and Ioc it. I know it's not the most secure thing in the world. However, it would interesting and provide a unique way to expand language support.

**Comment by jonathanindig** [comment]
Thank you for the kind words @aaronclong Much appreciated! 

We've been thinking about the best way to deal with adding new Interpreters / Kernels contributed by the community, but we haven't yet come to any conclusions about how we'd like to approach it, especially since we haven't yet gotten too much interest from anyone who wanted to contribute one! 

Similarly, we don't really have documentation here for much the same reason (it's a bit of a chicken and egg problem, I know). We'd be more than happy to help improve (ok, well, write) kernel documentation but we'd definitely need the help of a trailblazer from the community for this to really get done.  

Unfortunately given Polynote's language interop features, writing a new interpreter right now would require a non-trivial amount of work and Scala knowledge in addition to knowledge of the language you're trying to add. It'd be a bit more involved than implementing a JSON protocol on top of a REPL. (Definitely open to ideas on how to improve this however!) 

If you'd like to get started, check out some of the [other issues with the `language-support` tag](https://github.com/polynote/polynote/labels/language-support) and feel free to hop on to [our gitter](https://gitter.im/polynote/polynote).

**Comment by aaronclong** [comment]
@jonathanindig, my Scala ignorance is definitely a limiting factor in all this. However, maybe we can take a preemptive step and identifying the current abstractions in the codebase. It could be a good starting point to identify holes, things to leverage, and improvements that could be made. Undoubtedly, a simple JSON, GRPC, {insert serializable language-agnostic format} contract would be an awesome thing to consider in the future. I will look into the other language issues when I have time.

Off the bat, I am confused by the **Runtime** vs **Interpreter**  constructs that I saw in the code. Namely, it seems Java, Scala, and Spark have runtimes, Python doesn't. Python has an interpreter, which makes sense considering it is interpreted. However, there doesn't seem to be a common abstraction for language support. Could you clarify this at all for me? Likewise, I could just be reading the source code wrong.

https://github.com/polynote/polynote/tree/master/polynote-spark-runtime/src/main/scala/polynote/runtime/spark/reprs
https://github.com/polynote/polynote/tree/master/polynote-runtime/src
https://github.com/polynote/polynote/blob/master/polynote-kernel/src/main/scala/polynote/kernel/interpreter/python/PythonInterpreter.scala

**Comment by jeremyrsmith** [comment]
Hi @aaronclong,

`Interpreter` is the abstraction you're looking for. It's an interface (https://github.com/polynote/polynote/blob/master/polynote-kernel/src/main/scala/polynote/kernel/interpreter/Interpreter.scala) that's used for both Scala and Python (and would be the way to add a new language).

I think what @jonathanindig was describing as the challenging factor, is that interpreters all live together in the same kernel, and they all have to know how to share values with the JVM in order to accomplish that. For Scala this isn't a problem, and for Python we solved it with Jep. But other languages will also necessarily have a similar way to interoperate with the JVM via FFI (i.e. JNI or JNA).

**Comment by aaronclong** [comment]
Yeah, there would def need to work on that end. Most likely, it will be a long-term goal. However, I am curious about how I could dynamically load a **Third Party Implementation** of the **Interpreter** with custom dependency management? Is it already supported?

If I could do all this in an external library, the work could all fall back onto me in a custom library. 

As far as Swift is concerned, they have a **C** API, [SourceKit](https://github.com/apple/swift/tree/main/tools/SourceKit), that I could wrap in a JNI interface. From there, I could actually work on the kernel integration. 

I know these factors are particularly interesting to you @jeremyrsmith  or @jonathanindig. I am more trying to gauge whether the core project could support my **Frankenstein Experiment Kernel**, and what I would need to do for the core project to enable it.

**Comment by jonathanindig** [comment]
In theory, we already have the scaffolding in place (via [`ServiceLoaders`](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html)) that would support "drop-in" Interpreters (by "drop-in" I mean, stick `myCoolNewInterpreter.jar` in the right place Polynote will find and load it when it starts up). See the `Factory` and `Loader` traits [at the bottom of `Interpreter.scala`](https://github.com/polynote/polynote/blob/master/polynote-kernel/src/main/scala/polynote/kernel/interpreter/Interpreter.scala#L63-L104). 

In practice, since we haven't actually tried this before, I'm certain there are missing pieces to the end-to-end experience (things we haven't exposed, edge cases.. etc). We definitely want to get there though.

However, I'm not too concerned about Polynote's support of drop-in third-party interpreters, since that's not likely to be the bottleneck here. The tricky part would be writing the interpreter in the first place 😄 

So I'd recommend getting started directly in a fork of the core project without worrying about how the third party bit would work (copy the Python Interpreter or something). Once there's a working prototype, I expect it'd be pretty straightforward to extract that code out into a third party library and in the process finding all the bugs the our current loader implementation.

**Comment by aaronclong** [comment]
@jonathanindig sounds like a plan. It may take a while but don't doubt me 🙂

**Comment by jeremyrsmith** [comment]
@aaronclong let me know if you want to collab on this at some point. I actually really want to have some example of a plugin-based language interpreter (especially one that needs interesting stuff, like JNI-based interop and its own dependency ecosystem – this will expose a lot of things that we haven't thought through enough internally). I think a collaboration is going to be the best way to go on the first such (earnest) effort, because it will help me realize all the things that need fixing or changing (or especially documenting) upstream in Polynote.
